### PR TITLE
Make Element tree root generic

### PR DESCRIPTION
**Pull request by goderbauer**
Part of https://github.com/flutter/flutter/issues/121573.

This change allows the root of the element tree to be any `Element` (not just a `RenderObjectElement`). For that, it introduces the `RootElementMixin`, which can be used to make any Element subclass the root of the element tree. It replaces the `RootRenderObjectElement`, which is now deprecated in favor of the new mixin. In addition to that, it renames the `renderViewElement` accessor on the WidgetsBinding to `rootElement` to remove the implied assumption that the root is a `RenderObjectElement`.

This is in preparation for further changes towards multi-view rendering, where the root of the element tree will not be backed by render objects, see https://docs.flutter.dev/go/multiple-views for the details of that.

### CODE REVIEW

**Comment by goderbauer** [comment]
The fix for the flutter_plugins check is https://github.com/flutter/packages/pull/3539.

File: packages/flutter/lib/src/widgets/framework.dart [hunk 1]
```diff
@@ -6323,14 +6323,28 @@ abstract class RenderObjectElement extends Element {
   }
 }
 
-/// The element at the root of the tree.
+/// Deprecated. Will be removed in a future version of Flutter.
```
Comment (gspencergoog) [inline packages/flutter/lib/src/widgets/framework.dart:6326]: You could also mention that it is no longer used by Flutter.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter/lib/src/widgets/framework.dart (hunk 1 above)
Comment (goderbauer) [reply packages/flutter/lib/src/widgets/framework.dart]: Added.

**Comment by goderbauer** [comment]
https://github.com/flutter/packages/pull/3539 has rolled into the framework in https://github.com/flutter/flutter/pull/123542.

**Comment by goderbauer** [comment]
customer_testing failures are (unrelated) flakes hopefully fixed with https://github.com/flutter/tests/pull/239.

**Pull request merged**